Changes to the working group: make leader hireable.

shamil-gadelshin opened the pull request:
The working group leader should be hirable like a regular worker. All working group operations should be applicable to the leader. Only the council is allowed to invoke operations with the leader.

### Changes
- hiring workflow
- roles operations
- stake management
- improve comments
- stake refunding
- move out content working group structs to the separate file in the runtime
- optional slashing on the role termination
- changed get_worker_ids() (excluded the leader and renamed)
- new extrinsic added (update_reward_amount)

shamil-gadelshin committed:
Make the lead hirable in the working group module.

shamil-gadelshin committed:
Introduce HiringWorkflow for testing in the working group module.

shamil-gadelshin committed:
Add check for single leader hiring in the working group module.

shamil-gadelshin committed:
Implement terminate leader ability in the working group module.

shamil-gadelshin committed:
Filter the leader worker id from the get_regular_worker_ids() funtion.

shamil-gadelshin committed:
Add slashing during the terminate_role() in the working group module.

shamil-gadelshin committed:
Add stake refunding to the working group module.

shamil-gadelshin committed:
Add working group stake handler to the runtime.

shamil-gadelshin committed:
Refactor the runtime: move content working group code to the separate file.

shamil-gadelshin committed:
Unset lead on leaving the leader role in the working group.

shamil-gadelshin committed:
Modify update_role_account extrinsic in the workin group module.

shamil-gadelshin committed:
Modifey working group module to support leader operations.

Changes:
- modify recurring rewards module - make fields public to support testing
- add RequireSignedOrigin to the working group module errors
- change update_reward_amount authorization
- change decrease_stake authorization
- add tests

shamil-gadelshin committed:
Update comments in the working group

- small renaming
- update runtime comments

shamil-gadelshin committed:
Bump runtime version

shamil-gadelshin commented:
I made a refactoring in the `runtime/lib.rs`. Moved out `content working group` structs to the separate file and plan to move all structs of the `versioned store` module.

`runtime/lib.rs` is 1K lines of code. It is a very complex and concept rich file.
Could we introduce a rule - "Don't place the integration code in the runtime/lib.rs"? @mnaamani @bedeho

shamil-gadelshin committed:
Refactor fill_opening() extrinsic in the working group module.

bedeho commented on runtime-modules/working-group/src/errors.rs:
```diff
@@ -14,6 +14,12 @@ decl_error! {
+        CannotHireLeaderWhenLeaderExists,
+
+        /// Cannot fill opening with several applications.
+        CannotHireSeveralLeader,
```
CannotHireMultipleLeaders

shamil-gadelshin replied:
Ok

bedeho reviewed (changes requested):
- A heroic effort on all these updates, not easy to review though. I think it would have been better to separate out the name changes in a separate PR from other substntial conceptual changes, any way you could separate the two now?

- I don't think the representation of the lead vs. worker is quite right, fragments such as this in `update_role_account` are a telltale sign

```Rust
            // Update role account
            WorkerById::<T, I>::mutate(worker_id, |worker| {
                worker.role_account = new_role_account_id.clone()
            });

            // Update lead data if it is necessary.
            let lead = <CurrentLead::<T, I>>::get();
            if let Some(lead) = lead {
                if lead.worker_id == worker_id {
                    let new_lead = Lead{
                        role_account_id: new_role_account_id.clone(),
                        ..lead
                    };
                    <CurrentLead::<T, I>>::put(new_lead);
                }
            }
```

We are basically having to go and keep two separate but duplicated state fields in synch.
Or the fact that when you check that a worker is the lead, you do not even need to consult whether `CurrentLead` is actually set.

Initially, I was going to say that we should drop `Lead` type and make `CurrentLead` just hold a member id, but now that I think of it, why do we even need to hold on to any internal representation of who the current lead is in the module at all? We could just drop both? Let me know if I am lost at sea here, if not, lets do this change before I review any further.

shamil-gadelshin committed:
Refactor working group: fix error name

shamil-gadelshin committed:
Change CurrentLead concept in the working group module.

- remove Lead
- set current lead as worker id
- rename role_account to role_account_id
- fix tests

shamil-gadelshin commented:
I tried to have ID cache for the leader, but I agree with you that it creates more problems than solves. I change it to the worker_id.

shamil-gadelshin committed:
Add alias for working group module Application type.

bedeho commented on runtime-modules/working-group/src/lib.rs:
```diff
@@ -309,6 +311,11 @@ decl_storage! {
+
+        /// Map member id by hiring application id.
+        /// Required by StakingEventsHandler callback call to refund the balance on unstaking.
+        pub MemberIdByHiringApplicationId get(fn member_id_by_hiring_application_id):
```
Why is this needed? It was not needed in the content working group?

The only place I see it used is in `refund_working_group_stake`, in order to do this

```
let member_id = Module::<T, I>::member_id_by_hiring_application_id(hiring_application_id);
```

But you could already resolve the member id from an application by just looking at `Application::member_id`?

shamil-gadelshin replied:
The similar functionality in the content working group is here:

`pub UnstakerByStakeId get(unstaker_by_stake_id) config(): linked_map StakeId<T> => WorkingGroupUnstaker<LeadId<T>, CuratorId<T>>;`

But we have no `stake_id` in the working group. It is hidden inside the hiring module.

The method like this can be found in the runtime crate implemented within `ContentWorkingGroupStakingEventHandler`.

bedeho replied:
I think I understand.

bedeho commented on runtime-modules/working-group/src/lib.rs:
```diff
@@ -1259,6 +1187,101 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
+    pub fn get_regular_worker_ids() -> Vec<WorkerId<T>> {
+        let lead_worker_id = Self::current_lead();
+
+        <WorkerById<T, I>>::enumerate()
```
Do we have some limit somewhere which limits how large this map?

shamil-gadelshin replied:
Not yet, but we remove old workers on leaving or termination, so this collection should not be too long.

bedeho replied:
It's ultimately under the control of the lead, so they may choose to make it arbitrary long, which would be a problem.

So long as we are doing some enumeration, I think we will then need to have a hardcoded trait parameter which limits how many workers can be added to the map. I suggest such a check should be added to `fill_opening`.

bedeho commented on runtime-modules/working-group/src/lib.rs:
```diff
@@ -1259,6 +1187,101 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
+    }
+
+    /// Returns all existing worker id list excluding the current leader worker id.
+    pub fn get_regular_worker_ids() -> Vec<WorkerId<T>> {
```
You could make this slightly more flexible, and have it be `get_worker_ids(include_lead: bool)`.

shamil-gadelshin replied:
I prefer to leave it as it is because the only usage now if for assigning random storage provider. This random provider id is picked using only worker ids. Within the context of the picking storage provider to add 'include_lead' flag looks confusing for me. If you think this is a good idea to have such a feature - I'd rather add separate method - `gel_all_worker_ids()`.

bedeho replied:
ok

bedeho reviewed (changes requested):
Looks good!

I would say that this still was too much for one PR, quite a few changes did seem orthogonal and could have been delivered separately.

shamil-gadelshin committed:
Add ‘MaxWorkerNumberLimit’ constant to the working group module.

shamil-gadelshin committed:
Move worker limit from add_opening to fill_opening in the working group module.

bedeho reviewed (approved):
LGTM, nice job.
